### ISSUE TITLE
Support defining target shorthands that just build other targets

### ISSUE THREAD

**Issue by rhd** [mesonbuild/meson#368]
Hi Jussi,

Is it possible to setup a custom target as a shorthand for something longer.  For example:

`ninja myapp` instead of `ninja path/to/my/deeply/nested/app`

**Comment by nirbheek** [comment]
There is no built-in support for it, but you can create a `custom_target()` that does nothing (just call `echo`) and set the `depends:` kwarg to that target.

**Comment by keszybz** [comment]
How is this supposed to work? I'm trying to add a custom 'html' target to build some extra docs:

```python3
custom_target('html',
              depends : html_pages,
              command : ['echo'])
```
→ `Missing keyword argument "output".`

```python3
custom_target('html',
              depends : html_pages,
              output : [],
              command : ['echo'])

```
```
ninja: error: build.ninja:4288: expected path
build : CUSTOM_COMMAND  | man/binfmt.d.html man/bootctl.html man/bootup....
      ^ near here
WARNING: Could not create compilation database.
```
This breaks subsequent ninja-build invocations too:
```
ninja: error: build.ninja:4288: expected path
build : CUSTOM_COMMAND  | man/binfmt.d.html man/bootctl.html man/bootup....
      ^ near here
```

```
custom_target('html',
              depends : html_pages,
              output : ['html'],
              command : ['echo'])

```
→ OK, this works, but is ugly.

Ideally, there'd be a way to provide a grouping target, similarly to .PHONY in make, without having to provide any fake commands or output.

For added bonus, it'd be really nice to turn such a target into an install target, i.e. in the example above, to be able to specify something like `ninja-build install-html` and have the files comprising that target installed  into their install locations (even though the have the implied `install : false` keyword arg).

**Comment by nirbheek** [comment]
@keszybz completely agree, this workaround is not something I would endorse as a replacement, merely a stop-gap solution in case you need this feature for something critical.

We should definitely add target groupings like you suggest. It's not even difficult to add; see `mesonbuild/backends/ninjabackend.py` and `mesonbuild/backends/vs2010backend.py`.

**Comment by jpakkane** [comment]
This can be accomplished with a run target.

    run_target('docs', command : ['echo', 'done'], depends : [dep1, dep2])

**Comment by rhd** [comment]
@keszybz Did you find `custom_target` worked better than `run_target`? I noticed you're using `custom_target` over in your `systemd` port.

**Comment by fedelibre** [comment]
@jpakkane The problem in your example is that the dependency *must* be a build target. It cannot be a custom target. I wonder why.. custom target do have output files.

Consider the following minimal example:

```
project('test')

custom_target('hello.bak',
  input : 'hello.txt',
  output : 'hello.bak',
  command : ['cp', '@INPUT@', '@OUTPUT@'])

run_target('copy',
  command : 'echo',
  depends : 'hello.bak')
```

When I run `meson build` I get:

    meson.build:8:0: ERROR:  Depends items must be build targets.

**Comment by jpakkane** [comment]
You need to pass the object returned by `custom_target` to `depends`. Meson is not a string typed language, so strings do not work.

**Comment by keszybz** [comment]
> Did you find custom_target worked better than run_target? I noticed you're using custom_target over in your systemd port.

We moved systemd to meson-0.46 today, and I was going over our `meson.build` files checking if any old work-arounds can be removed...

`run_target` fails with: `ERROR:  Depends items must be build targets.`, so we still need to use `custom_target` with `echo`.

**Comment by jpakkane** [comment]
This can now be done with run targets so closing.

**Comment by nirbheek** [comment]
Also, `alias_target()`, which will be available in 0.52: https://github.com/mesonbuild/meson/pull/5606
